File: TacticalToasterUNTARGH/src/controllers/UntarSpawnController.ts
import { ApplicationContext } from "@spt/context/ApplicationContext";
import { ContextVariableType } from "@spt/context/ContextVariableType";
import { IGetRaidConfigurationRequestData } from "@spt/models/eft/match/IGetRaidConfigurationRequestData";
import { ILogger } from "@spt/models/spt/utils/ILogger";
import { ConfigServer } from "@spt/servers/ConfigServer";
import { DatabaseService } from "@spt/services/DatabaseService";
import { StaticRouterModService } from "@spt/services/mod/staticRouter/StaticRouterModService";
import { HttpResponseUtil } from "@spt/utils/HttpResponseUtil";
import { JsonUtil } from "@spt/utils/JsonUtil";
import { RandomUtil } from "@spt/utils/RandomUtil";
import { container, inject, injectable } from "tsyringe";
import { IMainConfig } from '../models/IMainConfig';
import { IBossLocationSpawn } from "../../types/models/eft/common/ILocationBase";
import { IBotConfig } from "@spt/models/spt/config/IBotConfig";
import { ConfigTypes } from "@spt/models/enums/ConfigTypes";

import mainConfig = require("../../config/main.json");
import { UNTARLogger } from "../logger";

//const mainConfig: IMainConfig = require("../../config/main.json");

@injectable()
export class UntarSpawnController {
    constructor(
        @inject("StaticRouterModService") protected staticRouter: StaticRouterModService,
        @inject("JsonUtil") protected jsonUtil: JsonUtil,
        @inject("RandomUtil") protected randomUtil: RandomUtil,
        @inject("ConfigServer") protected configServer: ConfigServer,
        @inject("DatabaseService") protected databaseService: DatabaseService,
        @inject("WinstonLogger") protected wLogger: ILogger,
        @inject("HttpResponseUtil") protected httpResponse: HttpResponseUtil,
        @inject("UNTARLogger") protected logger: UNTARLogger
    ) {}

    public adjustAllUntarSpawns() {
        try {
            const dbService = container.resolve<DatabaseService>("DatabaseService");
            const tables = dbService.getTables();
            const botConfig = this.configServer.getConfig<IBotConfig>(ConfigTypes.BOT);

            for (const map in mainConfig.locations) {
                this.logger.info(`Adjusting UNTAR spawns for ${map}.`);

                const mapConfig = mainConfig.locations[map];

                if (!tables.locations[map]) {
                    this.logger.info(`No location data found for ${map}. Skipping UNTAR spawn adjustment.`);
                    return;
                }

                const spawns = tables.locations[map].base.BossLocationSpawn as IBossLocationSpawn[];

                tables.locations[map].base.BossLocationSpawn = spawns.filter(
                    x => x.BossName.search("untar") === -1
                );

                if (mainConfig.locations[map] && mainConfig.locations[map].enablePatrols) {
                    this.logger.info(`Enabling UNTAR patrols for ${map}.`);

                    let validZones: string[] = mapConfig.patrolZones;

                    for (let i = 0; i < mapConfig.patrolAmount; i++) {
                        const patrolSize = this.randomUtil.getInt(mapConfig.patrolMin, mapConfig.patrolMax);
                        const patrol: IBossLocationSpawn = this.generatePatrol(
                            patrolSize,
                            mapConfig.patrolChance
                        );

                        patrol.BossZone = this.randomUtil.getArrayValue(validZones);
                        validZones = validZones.filter(zone => zone !== patrol.BossZone);

                        if (validZones.length === 0) {
                            validZones = mapConfig.patrolZones;
                        }

                        patrol.Time = this.randomUtil.getInt(
                            mapConfig.patrolTimeMin,
                            mapConfig.patrolTimeMax
                        );

                        if (mainConfig.debug.spawnInstantlyAlawys) {
                            this.logger.info(`Instantly spawning UNTAR patrol for ${map}.`);
                            patrol.Time = -1;
                        }

                        tables.locations[map].base.BossLocationSpawn.push(patrol);

                        this.logger.info(`Added (${mapConfig.patrolChance}% chance) UNTAR patrol of size ${patrolSize} to ${map} in zone ${patrol.BossZone} with a spawn time of ${patrol.Time} seconds.`);

                        //if (this.randomUtil.getChance100(mainConfig.locations[map].patrolChance)) {
                        //const patrolAmount = this.randomUtil.getInt(
                        //    mainConfig.locations[map].patrolMin,
                        //    mainConfig.locations[map].patrolMax
                        //);
                        //}
                    }
                    
                }
            }
        }
        catch (error) {
            this.logger.error(`Error adjusting untar spawns: ${error}`);
            throw error;
        }
    }

    public async adjustUntarSpawns(url: string, info: any, sessionID: string, output: string) {

        try {
            const dbService = container.resolve<DatabaseService>("DatabaseService");
            const tables = dbService.getTables();
            const botConfig = this.configServer.getConfig<IBotConfig>(ConfigTypes.BOT);
            const appContext = container.resolve<ApplicationContext>("ApplicationContext");
            const match = appContext.getLatestValue(ContextVariableType.RAID_CONFIGURATION).getValue<IGetRaidConfigurationRequestData>();
            const map = match.location.toLowerCase();
            this.logger.info(`Adjusting UNTAR spawns for ${map}.`);

            if (!tables.locations[map]) {
                this.logger.info(`No location data found for ${map}. Skipping UNTAR spawn adjustment.`);
                return output;
            }
            const spawns = tables.locations[map].base.BossLocationSpawn as IBossLocationSpawn[];

            if (mainConfig.locations[map] && mainConfig.locations[map].enablePatrols) {
                this.logger.info(`Enabling UNTAR patrols for ${map}.`);

                const mapConfig = mainConfig.locations[map];

                tables.locations[map].base.BossLocationSpawn = spawns.filter(
                    x => x.BossName.search("untar") === -1
                );

                let validZones: string[] = mapConfig.patrolZones;

                for (let i = 0; i < mapConfig.patrolAmount; i++) {
                    const patrolSize = this.randomUtil.getInt(mapConfig.patrolMin, mapConfig.patrolMax);
                    const patrol: IBossLocationSpawn = this.generatePatrol(
                        patrolSize,
                        mapConfig.patrolChance
                    );

                    patrol.BossZone = this.randomUtil.getArrayValue(validZones);
                    validZones = validZones.filter(zone => zone !== patrol.BossZone);

                    if (validZones.length === 0) {
                        validZones = mapConfig.patrolZones;
                    }

                    patrol.Time = this.randomUtil.getInt(
                        mapConfig.patrolTimeMin,
                        mapConfig.patrolTimeMax
                    );

                    tables.locations[map].base.BossLocationSpawn.push(patrol);

                    this.logger.info(`Added (${mapConfig.patrolChance}% chance) UNTAR patrol of size ${patrolSize} to ${map} in zone ${patrol.BossZone} with a spawn time of ${patrol.Time} minutes.`);

                    //if (this.randomUtil.getChance100(mainConfig.locations[map].patrolChance)) {
                    //const patrolAmount = this.randomUtil.getInt(
                    //    mainConfig.locations[map].patrolMin,
                    //    mainConfig.locations[map].patrolMax
                    //);
                    //}
                }
                
            }
        }
        catch (error) {
            this.logger.error(`Error adjusting untar spawns: ${error}`);
            throw error;
        }
        return output;
    }

    private generatePatrol(patrolSize: number, chance: number): IBossLocationSpawn {
        let bossType = "bossuntarlead";
        let secondLeader = "";
        let followers = patrolSize - 1;

        this.logger.info(`Generating UNTAR patrol of size ${patrolSize}.`);

        if (patrolSize >= mainConfig.patrols.minOfficerSize) {
            if (this.randomUtil.getChance100(mainConfig.patrols.officerChance)) {
                this.logger.info("UNTAR patrol leader changed to Officer.");
                bossType = "bossuntarofficer";
            }

            if (patrolSize >= mainConfig.patrols.minSecondLeaderSize) {
                if (this.randomUtil.getChance100(mainConfig.patrols.secondLeaderChance)) {
                    this.logger.info("UNTAR patrol second leader added.");
                    secondLeader = "bossuntarlead";
                    followers = followers - 1;
                }
            }
        }

        const bossInfo: IBossLocationSpawn = {
            BossChance: chance,
            BossDifficult: "normal",
            BossEscortAmount: "1",
            BossEscortDifficult: "normal",
            BossEscortType: "followeruntar",
            BossName: bossType,
            BossPlayer: false,
            BossZone: "",
            ForceSpawn: false,
            IgnoreMaxBots: true,
            RandomTimeSpawn: false,
            SpawnMode: [
                "regular",
                "pve"
            ],
            Supports: [],
            Time: -1,
            TriggerId: "",
            TriggerName: ""
        }

        if (secondLeader !== "") {
            bossInfo.Supports.push({
                "BossEscortAmount": "1",
                "BossEscortDifficult": [ "normal" ],
                "BossEscortType": secondLeader
            });
        }

        bossInfo.Supports.push({
            "BossEscortAmount": followers.toString(),
            "BossEscortDifficult": [ "normal" ],
            "BossEscortType": "followeruntar"
        });

        return bossInfo;
    }

    /*dynamicRouterModService.registerDynamicRouter(
            "UNTARDifficultiesRouter",
            [
                {
                    url: "/singleplayer/settings/bot/difficulties",
                    action: async (url: string, info: IEmptyRequestData, sessionID: string, output: string): Promise<string> => {
                        return untarDBController.getBotDifficulties(url, info, sessionID, output);
                    }
                }
            ],
            ""
        );*/
}

File: TacticalToasterUNTARGH/Controllers/UntarSpawnController.cs
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.DI;
using SPTarkov.Server.Core.Helpers;
using SPTarkov.Server.Core.Models.Eft.Common;
using SPTarkov.Server.Core.Models.Enums;
using SPTarkov.Server.Core.Models.Spt.Config;
using SPTarkov.Server.Core.Models.Spt.Mod;
using SPTarkov.Server.Core.Models.Spt.Server;
using SPTarkov.Server.Core.Models.Utils;
using SPTarkov.Server.Core.Servers;
using SPTarkov.Server.Core.Services;
using SPTarkov.Server.Core.Utils;
using SPTarkov.Server.Core.Utils.Json;
using System;

namespace TacticalToasterUNTARGH.Controllers;

[Injectable(InjectionType.Singleton)]
public class UntarSpawnController(
    JsonUtil jsonUtil,
    RandomUtil randomUtil,
    ConfigController configController,
    DatabaseService databaseService,
    UNTARLogger logger,
    HttpResponseUtil httpResponse)
{
    private readonly JsonUtil _jsonUtil = jsonUtil;
    private readonly RandomUtil _randomUtil = randomUtil;
    private readonly DatabaseService _databaseService = databaseService;
    private readonly UNTARLogger _logger = logger;
    private readonly HttpResponseUtil _httpResponse = httpResponse;
    private readonly ConfigController _configController = configController;

    public void AdjustAllUntarSpawns()
    {
        try
        {
            var tables = _databaseService.GetTables();
            var mainConfig = _configController.ModConfig;

            foreach (var map in mainConfig.locations.Keys)
            {
                _logger.Info($"Adjusting UNTAR spawns for {map}.");

                if (!tables.Locations.GetDictionary().ContainsKey(map))
                {
                    _logger.Info($"No location data found for {map}. Skipping UNTAR spawn adjustment.");
                    continue;
                }

                var mapConfig = mainConfig.locations[map];
                var patrolConfig = mapConfig.patrol;
                var checkpointConfig = mapConfig.checkpoint;
                var spawns = tables.Locations.GetDictionary()[map].Base.BossLocationSpawn;

                // Remove existing UNTAR spawns
                spawns.RemoveAll(x => x.BossName.Contains("untar"));

                if (patrolConfig.enablePatrols)
                {
                    AdjustPatrolSpawnsForMap(map, mapConfig, mainConfig, spawns);
                }

                if (checkpointConfig.enableCheckpoints)
                {
                    AdjustCheckpointSpawnsForMap(map, mapConfig, mainConfig, spawns);
                }
            }
        }
        catch (Exception ex)
        {
            _logger.Error($"Error adjusting UNTAR spawns: {ex.Message}");
            throw;
        }
    }

    /*
    public string AdjustUntarSpawns(string url, object info, string sessionID, string output)
    {
        try
        {
            var tables = _databaseService.GetTables();
            var appContext = ApplicationContext.Instance;
            var match = appContext.GetLatestValue<IGetRaidConfigurationRequestData>(ContextVariableType.RAID_CONFIGURATION);
            var map = match.Location.ToLower();

            _logger.Info($"Adjusting UNTAR spawns for {map}.");

            if (!tables.Locations.ContainsKey(map))
            {
                _logger.Info($"No location data found for {map}. Skipping UNTAR spawn adjustment.");
                return output;
            }

            var spawns = tables.Locations[map].Base.BossLocationSpawn;

            if (MainConfig.Locations[map].EnablePatrols)
            {
                var mapConfig = MainConfig.Locations[map];
                spawns.RemoveAll(x => x.BossName.Contains("untar"));

                var validZones = new List<string>(mapConfig.PatrolZones);

                for (int i = 0; i < mapConfig.PatrolAmount; i++)
                {
                    var patrolSize = _randomUtil.GetInt(mapConfig.PatrolMin, mapConfig.PatrolMax);
                    var patrol = GeneratePatrol(patrolSize, mapConfig.PatrolChance);

                    patrol.BossZone = _randomUtil.GetArrayValue(validZones);
                    validZones.Remove(patrol.BossZone);

                    if (!validZones.Any())
                    {
                        validZones = new List<string>(mapConfig.PatrolZones);
                    }

                    patrol.Time = _randomUtil.GetInt(mapConfig.PatrolTimeMin, mapConfig.PatrolTimeMax);
                    spawns.Add(patrol);

                    _logger.Info($"Added ({mapConfig.PatrolChance}% chance) UNTAR patrol of size {patrolSize} to {map} in zone {patrol.BossZone} with a spawn time of {patrol.Time} minutes.");
                }
            }
        }
        catch (Exception ex)
        {
            _logger.Error($"Error adjusting UNTAR spawns: {ex.Message}");
            throw;
        }

        return output;
    }
    */

    private void AdjustPatrolSpawnsForMap(string map, MapConfig mapConfig, MainConfig mainConfig, List<BossLocationSpawn> spawns)
    {
        var patrolConfig = mapConfig.patrol;

        _logger.Info($"Enabling UNTAR patrols for {map}.");
        var validZones = new List<string>(patrolConfig.patrolZones);

        for (int i = 0; i < patrolConfig.patrolAmount; i++)
        {
            var patrolSize = _randomUtil.GetInt(patrolConfig.patrolMin, patrolConfig.patrolMax);
            var patrol = GeneratePatrol(patrolSize, mainConfig.debug.spawnAlways ? 100 : patrolConfig.patrolChance);

            patrol.BossZone = _randomUtil.GetArrayValue(validZones);
            validZones.Remove(patrol.BossZone);

            if (validZones.Count == 0)
            {
                validZones = new List<string>(patrolConfig.patrolZones);
            }

            patrol.Time = _randomUtil.GetInt(patrolConfig.patrolTimeMin, patrolConfig.patrolTimeMax);

            if (mainConfig.debug.spawnInstantlyAlways)
            {
                _logger.Info($"Instantly spawning UNTAR patrol for {map}.");
                patrol.Time = -1;
            }

            spawns.Add(patrol);

            _logger.Info($"Added ({patrolConfig.patrolChance}% chance) UNTAR patrol of size {patrolSize} to {map} in zone {patrol.BossZone} with a spawn time of {patrol.Time} seconds.");
        }
    }

    private void AdjustCheckpointSpawnsForMap(string map, MapConfig mapConfig, MainConfig mainConfig, List<BossLocationSpawn> spawns)
    {
        var checkpointConfig = mapConfig.checkpoint;

        _logger.Info($"Enabling UNTAR checkpoint for {map}.");
        var validZones = new List<ZoneCheckpointConfig>(checkpointConfig.checkpointZones);

        for (int i = 0; i < checkpointConfig.checkpointAmount; i++)
        {
            var checkpointZoneConfig = _randomUtil.GetArrayValue(validZones);
            validZones.Remove(checkpointZoneConfig);

            var patrolSize = _randomUtil.GetInt(checkpointZoneConfig.checkpointMin, checkpointZoneConfig.checkpointMax);
            var patrol = GeneratePatrol(patrolSize, mainConfig.debug.spawnAlways ? 100 : checkpointZoneConfig.checkpointChance, false);

            patrol.BossZone = checkpointZoneConfig.checkpointZone;

            if (validZones.Count == 0)
            {
                validZones = [.. checkpointConfig.checkpointZones];
            }

            patrol.Time = -1;//_randomUtil.GetInt(patrolConfig.patrolTimeMin, patrolConfig.patrolTimeMax);

            if (mainConfig.debug.spawnInstantlyAlways)
            {
                _logger.Info($"Instantly spawning UNTAR checkpoint for {map}.");
                patrol.Time = -1;
            }

            spawns.Add(patrol);

            _logger.Info($"Added ({checkpointZoneConfig.checkpointChance}% chance) UNTAR checkpoint of size {patrolSize} to {map} in zone {patrol.BossZone} with a spawn time of {patrol.Time} seconds.");
        }
    }

    private BossLocationSpawn GeneratePatrol(int patrolSize, float chance, bool isPatrol = true)
    {
        var bossType = "bossuntarlead";
        var secondLeader = string.Empty;
        var followers = patrolSize - 1;
        var mainConfig = _configController.ModConfig;
        var genConfig = mainConfig.patrols;

        if (isPatrol == false)
            genConfig = mainConfig.checkpoints;

        _logger.Info($"Generating UNTAR patrol of size {patrolSize}.");

        if (patrolSize >= genConfig.minOfficerSize && _randomUtil.GetChance100(genConfig.officerChance))
        {
            _logger.Info("UNTAR patrol leader changed to Officer.");
            bossType = "bossuntarofficer";
        }

        if (patrolSize >= genConfig.minSecondLeaderSize && _randomUtil.GetChance100(genConfig.secondLeaderChance))
        {
            _logger.Info("UNTAR patrol second leader added.");
            secondLeader = "bossuntarlead";
            followers--;
        }

        var supportsList = new List<BossSupport>();

        

        if (!string.IsNullOrEmpty(secondLeader))
        {
            supportsList.Add(new BossSupport
            {
                BossEscortAmount = "1",
                BossEscortDifficulty = new ListOrT<string>(["normal"], null),
                BossEscortType = secondLeader
            });
        }

        supportsList.Add(new BossSupport
        {
            BossEscortAmount = followers.ToString(),
            BossEscortDifficulty = new ListOrT<string>(["normal"], null),
            BossEscortType = "followeruntar"
        });

        var bossInfo = new BossLocationSpawn
        {
            BossChance = chance,
            BossDifficulty = "normal",
            BossEscortAmount = "1",
            BossEscortDifficulty = "normal",
            BossEscortType = "followeruntar",
            BossName = bossType,
            IsBossPlayer = false,
            BossZone = string.Empty,
            ForceSpawn = false,
            IgnoreMaxBots = true,
            IsRandomTimeSpawn = false,
            SpawnMode = new[] { "regular", "pve" },
            Supports = supportsList,
            Time = -1,
            TriggerId = string.Empty,
            TriggerName = string.Empty
        };

        return bossInfo;
    }
}